tests/test_bookmark.py

```diff
@@ -0,0 +1,220 @@
+
+                # collect information for assertions from syncs 1 & 2 base on expected values
+                first_sync_count = first_sync_record_count.get(stream, 0)
+                second_sync_count = second_sync_record_count.get(stream, 0)
+                first_sync_messages = [record.get('data') for record in
+                                       first_sync_records.get(
+                                           stream, {}).get('messages', [])
+                                       if record.get('action') == 'upsert']
+                second_sync_messages = [record.get('data') for record in
+                                        second_sync_records.get(
+                                            stream, {}).get('messages', [])
+                                        if record.get('action') == 'upsert']
+                first_bookmark_key_value = first_sync_bookmarks.get('bookmarks', {stream: None}).get(stream)
+                second_bookmark_key_value = second_sync_bookmarks.get('bookmarks', {stream: None}).get(stream)  
+
+                if expected_replication_method == self.INCREMENTAL :
+                    # collect information specific to incremental streams from syncs 1 & 2
+                    replication_key = next(iter(expected_replication_keys[stream]))
+                    if stream in ['product_reviews','order_fulfillments','product_variants'] :
+                        first_bookmark_value_utc = {key : self.convert_state_to_utc(value) for key, value in first_bookmark_key_value.items()}
+                        second_bookmark_value_utc = {key : self.convert_state_to_utc(value) for key, value in second_bookmark_key_value.items()}
+                        simulated_bookmark_value = { key : self.convert_state_to_utc(value) for key, value in new_states['bookmarks'][stream].items()}
+
+                        simulated_bookmark_minus_lookback = simulated_bookmark_value 
+
+                        # Verify the first sync sets a bookmark of the expected form
+                        self.assertIsNotNone(first_bookmark_key_value)
+
+                        # Verify the second sync sets a bookmark of the expected form
+                        self.assertIsNotNone(second_bookmark_key_value)
+
+                        # Verify the second sync bookmark is Equal to the first sync bookmark
+                        # assumes no changes to data during test
+                        for item in first_bookmark_key_value.keys():
+                            self.assertGreaterEqual(second_bookmark_key_value.get(item), first_bookmark_key_value.get(item))
+
+                        # Bookmark is in the format of {bookmark_keys1 : date1,bookmark_keys2 : date2,.....}
+                        bookmark_keys = {
+                            "product_reviews": "product_yotpo_id",
+                            "order_fulfillments": "order_id",
+                            "product_variants": "yotpo_product_id"} 
```

Comment:
These 3 streams product_reviews,product_variants, and  order_fulfillments are child streams of product and order streams, respectively. 
The bookmarking is maintained for each item( i.e. product id / order id) of the parent stream, while the DateTime value 'updated_at'/ 'created_at' is retrieved from the records (which is the actual replication key)
e.g. We are storing the state file for order_fulfillments then it will be a mapping of multiple order ids with their respective replication key values.
{order_id1: update_at1, order_id2: updated_at2, order_id3: updated_at3,.........,etc}